README.md
# onyx-peer-http-query

Onyx Peer HTTP Query provides an inbuilt HTTP server to service replica and
cluster queries that can be directed at Onyx nodes. One use case is to provide
a health check for your Onyx nodes, as it becomes easy to determine what a
node's view of the cluster is.

### HTTP Server

This library exposes an HTTP server to service replica and cluster queries across languages.

To use it, add onyx-peer-http-query to your dependencies:
```
[org.onyxplatform/onyx-peer-http-query "0.10.0.0-beta13"]
```

Require onyx.http-query in your peer bootup namespace:

```clojure
(:require [onyx.http-query])
```


And add the following lines to Onyx's peer-config
```
 :onyx.query/server? true
 :onyx.query.server/port 8080
```

In addition, you can optionally add the IP to listen on with
```
 :onyx.query.server/ip "127.0.0.1"
```

JMX selectors can be whitelisted/queried via the peer-config:
e.g.
```
 :onyx.query.server/metrics-selectors ["org.onyxplatform:*" "com.amazonaws.management:*"]
```

The default behaviour is
```
 :onyx.query.server/metrics-selectors ["*:*"]
```

Individual metrics tags can be blacklisted via the peer-config:
```
 :onyx.query.server/metrics-blacklist [#"blacklisted_tag1" #"blacklistregex.*"]
```

### Accessing the HTTP server

Then query it to get a view of that nodes understanding of the cluster:

```
$ http --json http://localhost:8080/replica/peers
```

```json
HTTP/1.1 200 OK
Content-Length: 197
Content-Type: application/json
Date: Tue, 23 Feb 2016 03:35:08 GMT
Server: Jetty(9.2.10.v20150310)

{
    "as-of-entry": 12,
    "as-of-timestamp": 1456108757818,
    "result": [
        "e52df81d-38c9-44e6-9e3d-177d3e83292b",
        "fd4725f9-3429-49eb-840d-6c3e29cecc41",
        "fc933dda-7260-4547-93fc-241a02ca599a"
    ],
    "status": "success"
}
```

Note `as-of-entry` and `as-of-timestamp`. By comparing `as-of-entry` between
nodes, you can discover whether a node is lagging behind the cluster.

Further API endpoints are [described here](doc/server-api.md).

### Endpoints 

The Replica Query Server has a number of endpoints for accessing the
information about a running Onyx cluster. Below we display the HTTP method, the
URI, the docstring for the route, and any associated parameters that it takes
in its query string.

#### Summary

- `/health`
- `/peergroup/heartbeat`
- `/peergroup/health`
- `/network/media-driver/active`
- `/metrics`
- `/job/catalog`
- `/job/flow-conditions`
- `/job/lifecycles`
- `/job/task`
- `/job/triggers`
- `/job/windows`
- `/job/workflow`
- `/job/exception`
- `/replica`
- `/replica/completed-jobs`
- `/replica/job-allocations`
- `/replica/job-scheduler`
- `/replica/jobs`
- `/replica/killed-jobs`
- `/replica/peer-site`
- `/replica/peer-state`
- `/replica/peers`
- `/replica/task-allocations`
- `/replica/task-scheduler`
- `/replica/tasks`

---

---

##### Route

`[:get]` `/health`

##### Query Params Schema

`{"threshold" java.lang.Long}`

##### Docstring

A single health check call to check whether the following statuses are healthy: `/network/media-driver/active`, `/peergroup/heartbeat`. Considers the peer group dead if timeout is greater than ?threshould=VALUE. Returns status 200 if healthy, 500 if unhealthy. Use this route for failure monitoring, automatic rebooting, etc.

---

##### Route

`[:get]` `/peergroup/heartbeat`


##### Query Params Schema

`{}`

##### Docstring

Returns the number of milliseconds since the peer group last heartbeated.

---

##### Route

`[:get]` `/peergroup/health`


##### Query Params Schema

`{}`

##### Docstring

A single health check call to check whether the peer group has heartbeated more recently than a threshold. Considers the peer group dead if timeout is greater than ?threshould=VALUE. Returns status 200 if healthy, 500 if unhealthy. Use this route for failure monitoring, automatic rebooting, etc.


##### Route

`[:get]` `/network/media-driver`


##### Query Params Schema

`{}`

##### Docstring

Returns a map describing the media driver status.
e.g.
```clojure
{:active true, 
 :driver-timeout-ms 10000, 
 :log "INFO: Aeron directory /var/folders/c5/2t4q99_53mz_c1h9hk12gn7h0000gn/T/aeron-lucas exists
       INFO: Aeron CnC file /var/folders/c5/2t4q99_53mz_c1h9hk12gn7h0000gn/T/aeron-lucas/cnc.dat exists
       INFO: Aeron toDriver consumer heartbeat is 687 ms old"}
```

---

##### Route

`[:get]` `/network/media-driver/active`


##### Query Params Schema

`{}`

##### Docstring

Returns a boolean for whether the media driver is active and has heartbeated within driver-timeout-ms milliseconds.

---

##### Route

`[:get]` `/metrics`


##### Query Params Schema

`{}`

##### Docstring

Returns any numeric JMX metrics contained in this VM, converted to prometheus tags.

---



##### Route

`[:get]` `/job/catalog`


##### Query Params Schema

`{"job-id" java.lang.String}`

##### Docstring

Given a job id, returns catalog for this job.

---

##### Route

`[:get]` `/job/flow-conditions`


##### Query Params Schema

`{"job-id" java.lang.String}`

##### Docstring

Given a job id, returns flow conditions for this job.

---

##### Route

`[:get]` `/job/lifecycles`


##### Query Params Schema

`{"job-id" java.lang.String}`

##### Docstring

Given a job id, returns lifecycles for this job.

---

##### Route

`[:get]` `/job/task`


##### Query Params Schema

`{"job-id" java.lang.String, "task-id" java.lang.String}`

##### Docstring

Given a job id and task id, returns catalog entry for this task.

---

##### Route

`[:get]` `/job/triggers`


##### Query Params Schema

`{"job-id" java.lang.String}`

##### Docstring

Given a job id, returns triggers for this job.

---

##### Route

`[:get]` `/job/windows`


##### Query Params Schema

`{"job-id" java.lang.String}`

##### Docstring

Given a job id, returns windows for this job.

---

##### Route

`[:get]` `/job/workflow`


##### Query Params Schema

`{"job-id" java.lang.String}`

##### Docstring

Given a job id, returns workflow for this job.

---

##### Route

`[:get]` `/job/exception`


##### Query Params Schema

`{"job-id" java.lang.String}`

##### Docstring

Given a job id, returns the exception that killed this job, if one exists.

---

##### Route

`[:get]` `/replica`


##### Query Params Schema

``

##### Docstring

Derefences the replica as an immutable value.

---

##### Route

`[:get]` `/replica/completed-jobs`


##### Query Params Schema

``

##### Docstring

Lists all the job ids that have been completed.

---

##### Route

`[:get]` `/replica/job-allocations`


##### Query Params Schema

``

##### Docstring

Returns a map of job id -> task id -> peer ids, denoting which peers are assigned to which tasks.

---

##### Route

`[:get]` `/replica/job-scheduler`


##### Query Params Schema

##### Docstring

Returns the job scheduler for this tenancy of the cluster.

---

##### Route

`[:get]` `/replica/jobs`


##### Query Params Schema

``

##### Docstring

Lists all non-killed, non-completed job ids.

---

##### Route

`[:get]` `/replica/killed-jobs`


##### Query Params Schema

``

##### Docstring

Lists all the job ids that have been killed.

---

##### Route

`[:get]` `/replica/peer-site`


##### Query Params Schema

`{"peer-id" java.lang.String}`

##### Docstring

Given a peer id, returns the Aeron hostname and port that this peer advertises to the rest of the cluster.

---

##### Route

`[:get]` `/replica/peer-state`


##### Query Params Schema

`{"peer-id" java.lang.String}`

##### Docstring

Given a peer id, returns its current execution state (e.g. :idle, :active, etc).

---

##### Route

`[:get]` `/replica/peers`


##### Query Params Schema

``

##### Docstring

Lists all the peer ids.

---

##### Route

`[:get]` `/replica/task-allocations`


##### Query Params Schema

``

##### Docstring

Given a job id, returns a map of task id -> peer ids, 
  denoting which peers are assigned to which tasks for this job only.

---

##### Route

`[:get]` `/replica/task-scheduler`


##### Query Params Schema

`{"job-id" java.lang.String}`

##### Docstring

Given a job id, returns the task scheduler for this job.

---

##### Route

`[:get]` `/replica/tasks`


##### Query Params Schema

`{"job-id" java.lang.String}`

##### Docstring

Given a job id, returns all the task ids for this job.


## License

Copyright © 2016 Distributed Masonry Inc.

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGES.MD
#### 0.10.0.0-beta12
* Added the following health check endpoints.
- `/health`
- `/peergroup/heartbeat`
- `/peergroup/health`

#### 0.9.15.0
* Added new network (Aeron Media Driver) health check endpoint at "/network/media-driver" (returns a map) and "/network/media-driver/active" (returns a boolean).


test/onyx/health_test.clj
(ns onyx.health-test
  (:require [clojure.core.async :refer [chan >!! <!! close! sliding-buffer]]
            [clojure.test :refer [deftest is testing]]
            [onyx.plugin.core-async :refer [take-segments!]]
            [onyx.test-helper :refer [load-config with-test-env add-test-env-peers!]]
            [onyx.static.uuid :refer [random-uuid]]
            [com.stuartsierra.component :as component]
            [onyx.http-query]
            [clj-http.client :as client]
            [onyx.api]))

(def n-messages 100)

(def in-chan (atom nil))
(def in-buffer (atom nil))

(def out-chan (atom nil))

(defn inject-in-ch [event lifecycle]
  {:core.async/buffer in-buffer
   :core.async/chan @in-chan})

(defn inject-out-ch [event lifecycle]
  {:core.async/chan @out-chan})

(def in-calls
  {:lifecycle/before-task-start inject-in-ch})

(def out-calls
  {:lifecycle/before-task-start inject-out-ch})

(defn my-inc [{:keys [n] :as segment}]
  (assoc segment :n (inc n)))

(deftest health-test
  (let [id (random-uuid)
        env-config {:zookeeper/address "127.0.0.1:2188"
                    :zookeeper/server? true
                    :zookeeper.server/port 2188
                    :onyx.bookkeeper/server? true
                    :onyx.bookkeeper/delete-server-data? true
                    :onyx.bookkeeper/local-quorum? true
                    :onyx.bookkeeper/local-quorum-ports [3196 3197 3198]
                    :onyx/tenancy-id id} 
        peer-config {:zookeeper/address "127.0.0.1:2188"
                     :onyx.peer/job-scheduler :onyx.job-scheduler/greedy
                     :onyx.peer/zookeeper-timeout 60000
                     :onyx.messaging.aeron/embedded-driver? true
                     :onyx.messaging/allow-short-circuit? false
                     :onyx.messaging/impl :aeron
                     :onyx.messaging/peer-port 40199
                     :onyx.messaging/bind-addr "localhost"
                     :onyx.query.server/metrics-selectors ["com.amazonaws.management:*" "*:*"]
                     :onyx/tenancy-id id
                     :onyx.query/server? true
                     :onyx.query.server/ip "127.0.0.1"
                     :onyx.query.server/port 8091}]
    (with-test-env [test-env [3 env-config peer-config]]
      (let [batch-size 20
	    catalog [{:onyx/name :in
		      :onyx/plugin :onyx.plugin.core-async/input
		      :onyx/type :input
		      :onyx/medium :core.async
		      :onyx/batch-size batch-size
		      :onyx/max-peers 1
		      :onyx/doc "Reads segments from a core.async channel"}

		     {:onyx/name :inc
		      :onyx/fn :onyx.health-test/my-inc
		      :onyx/type :function
		      :onyx/batch-size batch-size}

		     {:onyx/name :out
		      :onyx/plugin :onyx.plugin.core-async/output
		      :onyx/type :output
		      :onyx/medium :core.async
		      :onyx/batch-size batch-size
		      :onyx/max-peers 1
		      :onyx/doc "Writes segments to a core.async channel"}]
	    workflow [[:in :inc] [:inc :out]]
	    lifecycles [{:lifecycle/task :in
			 :lifecycle/calls :onyx.health-test/in-calls}
			{:lifecycle/task :out
			 :lifecycle/calls :onyx.health-test/out-calls}]
	    _ (reset! in-chan (chan (inc n-messages)))
	    _ (reset! in-buffer {})
	    _ (reset! out-chan (chan (sliding-buffer (inc n-messages))))
	    _ (doseq [n (range n-messages)]
		(>!! @in-chan {:n n}))
	    job-id (:job-id (onyx.api/submit-job peer-config
						 {:catalog catalog
						  :workflow workflow
						  :lifecycles lifecycles
						  :task-scheduler :onyx.task-scheduler/balanced
						  :metadata {:job-name :click-stream}}))
            _ (Thread/sleep 1000)
	    peers (:result (clojure.edn/read-string (:body (client/get "http://127.0.0.1:8091/replica/peers"))))]
        (mapv (fn [[{:keys [uri]} {:keys [query-params-schema]}]]
                (println "uri:" uri)
                (if (= "/metrics" uri)
                  (do
                  (println (:body (client/get (str "http://127.0.0.1:8091" uri) 
                                                  {:query-params {}})))
                   (is (re-find #"replica_version" 
                               (:body (client/get (str "http://127.0.0.1:8091" uri) 
                                                  {:query-params {}})))))
                  (is (= :success 
                         (:status 
                          (doto 
                            (clojure.edn/read-string 
                             (:body (client/get (str "http://127.0.0.1:8091" uri) 
                                                {:query-params {"task-id" "out"
                                                                "threshold" 10000
                                                                "peer-id" (first peers)
                                                                "job-id" (str job-id)}})))
                            println)))))) 
              onyx.http-query/endpoints)
        (let [_ (close! @in-chan)
              _ (onyx.test-helper/feedback-exception! peer-config job-id)
              results (take-segments! @out-chan 500)]
          (is (= [job-id] (:result (clojure.edn/read-string (:body (client/get "http://127.0.0.1:8091/replica/completed-jobs"))))))
          (let [expected (set (map (fn [x] {:n (inc x)}) (range n-messages)))]
            (is (= expected (set results)))))))))


test/onyx_health/core_test.clj
(ns onyx-health.core-test
  (:require [clojure.test :refer :all]
            [onyx-health.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


project.clj
(defproject org.onyxplatform/onyx-peer-http-query "0.10.0.0-SNAPSHOT"
  :description "An Onyx health and query HTTP server"
  :url "https://github.com/onyx-platform/onyx-peer-http-query"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[org.clojure/clojure "1.8.0"]
                 ^{:voom {:repo "git@github.com:onyx-platform/onyx.git" :branch "master"}}
                 [org.onyxplatform/onyx "0.10.0-20170428_035232-g0d0a9e3"]
                 [ring/ring-core "1.5.1"]
                 [org.clojure/java.jmx "0.3.4"]
                 [ring-jetty-component "0.3.1"]
                 [cheshire "5.7.0"]]
  :repositories {"snapshots" {:url "https://clojars.org/repo"
                              :username :env
                              :password :env
                              :sign-releases false}
                 "releases" {:url "https://clojars.org/repo"
                             :username :env
                             :password :env
                             :sign-releases false}}
  :profiles {:dev {:dependencies [[clj-http "3.4.1"]]
                   :plugins [[lein-set-version "0.4.1"]
                             [lein-update-dependency "0.1.2"]
                             [lein-pprint "1.1.1"]]}})


src/onyx/metrics_endpoint.clj
(ns onyx.metrics-endpoint
  (:require [clojure.string]
            [clojure.java.jmx :as jmx]))

(def extractions
  {:job-id #"(?:[.]|=)job[.]([^_]+)"
   :task #"(?:[.]|=)task[.]([^_]+)"
   :slot-id #"(?:[.]|=)slot-id[.]([^_]+)"
   :peer-id #"(?:[.]|=)peer-id[.]([^_]+)"})

(defn remove-tags [s]
  (clojure.string/replace (reduce (fn [s [k v]]
                                    (clojure.string/replace s v ""))
                                  s
                                  extractions) 
                          #"_$"
                          ""))

(defn canonicalize [s]
  (clojure.string/replace s #"[^a-zA-Z0-9:_]" "_"))

(defn remove-jmx-prefix [s]
  (-> s
      (clojure.string/replace #"^name=" "")
      (clojure.string/replace #"^type=" "")))

(defn extract-metric [s]
  (loop [[p & ps] (clojure.string/split s #"[.]")
         tags []
         metric ""]
    (cond (nil? p)
          {:tags tags :metric metric}

          (#{"task" "job" "peer-id" "slot-id"} p) 
          (recur (rest ps)
                 (conj tags p (first ps))
                 metric)
          :else
          (recur ps
                 tags
                 (if (empty? metric)
                   p
                   (str metric "_" p))))))

(defn job-metric->metric-str [metric-str attribute]
  (let [{:keys [tags metric]} (extract-metric (remove-jmx-prefix metric-str))
        tag-str (->> tags 
                     (partition 2)
                     (map (fn [[name value]]
                            (format "%s=\"%s\"" (canonicalize name) (canonicalize value))))
                     (clojure.string/join ", ")
                     (format "{%s}"))] 
    (format "%s_%s%s" 
            (canonicalize metric)
            (canonicalize (name attribute)) 
            (if (= "{}" tag-str) "" tag-str))))

(defn blacklisted? [blacklists metric]
  (boolean 
   (some (fn [r] (re-find r metric)) 
         blacklists)))

(defn metrics-endpoint [peer-config]
  (let [builder (java.lang.StringBuilder.)
        blacklists (:onyx.query.server/metrics-blacklist peer-config)
        mbean-selectors (or (:onyx.query.server/metrics-selectors peer-config) ["*:*"])
        selected (atom #{})]
    (doseq [selector mbean-selectors]
      (doseq [mbean (jmx/mbean-names selector)] 
        (try 
         (let [canonical-key (.getCanonicalKeyPropertyListString mbean)]
           ;; deduplicate
           (when-not (@selected canonical-key)
             (swap! selected conj canonical-key)
             (doseq [attribute (jmx/attribute-names mbean)]
               (try 
                (let [value (jmx/read mbean attribute)] 
                  (cond (number? value) 
                        (let [metric (job-metric->metric-str (.getCanonicalKeyPropertyListString mbean) attribute)]
                          (when-not (blacklisted? blacklists metric)
                            (.append builder (format "%s %s" metric value))
                            (.append builder "\n")))
                        (map? value) 
                        (run! (fn [[k v]]
                                (when (number? v)
                                  (let [metric (job-metric->metric-str (.getCanonicalKeyPropertyListString mbean) 
                                                                       (str (name attribute) "_" (name k)))]
                                    (when-not (blacklisted? blacklists metric)
                                      (.append builder (format "%s %s" metric v))
                                      (.append builder "\n")))))
                              value)))
                ;; Safe to swallow, jobs and beans are ephemeral
                (catch javax.management.RuntimeMBeanException _)))))
         ;; Safe to swallow, jobs and beans are ephemeral
         (catch javax.management.InstanceNotFoundException _))))
    (str builder)))


src/onyx/peer_query/job_query.clj
(ns onyx.peer-query.job-query
  (:require [onyx.log.commands.common :as common]
            [onyx.log.zookeeper :as zk]
            [onyx.static.planning :refer [find-task]]
            [onyx.peer-query.replica-query :as rq]
            [onyx.extensions :as extensions])
  (:import [org.apache.zookeeper.KeeperException$NoNodeException]))

(defn ^{:no-doc true} get-log [log]
  (:log (:env log)))

(defn ^{:no-doc true} find-in-seq [coll k v]
  (first (filter #(= v (k %)) coll)))

(defn workflow
  "Given a job id, returns workflow for this job."
  [log job-id]
  (extensions/read-chunk log :workflow job-id))

(defn catalog
  "Given a job id, returns catalog for this job."
  [log job-id]
  (extensions/read-chunk log :catalog job-id))

(defn job-metadata
  "Given a job id, returns job-metadata for this job."
  [log job-id]
  (or (extensions/read-chunk log :job-metadata job-id)))

(defn flow-conditions
  "Given a job id, returns flow conditions for this job."
  [log job-id]
  (or (extensions/read-chunk log :flow-conditions job-id) []))

(defn lifecycles
  "Given a job id, returns lifecycles for this job."
  [log job-id]
  (or (extensions/read-chunk log :lifecycles job-id) []))

(defn windows
  "Given a job id, returns windows for this job."
  [log job-id]
  (or (extensions/read-chunk log :windows job-id) []))

(defn triggers
  "Given a job id, returns triggers for this job."
  [log job-id]
  (or (extensions/read-chunk log :triggers job-id) []))

(defn exception
  "Given a job id, return the exception for this job, if any"
  [log job-id]
  (try (extensions/read-chunk log :exception job-id)
       ;; Workaround for the fact that we don't handle this upstream
       (catch org.apache.zookeeper.KeeperException$NoNodeException nne
         (println "Exception not found for job" job-id))))

(defn job-information 
  "Given a job id and task id, returns job data for this task."
  [log replica job-id]
  {:workflow (workflow log job-id)
   :catalog (catalog log job-id)
   :flow-conditions (flow-conditions log job-id)
   :lifecycles (lifecycles log job-id)
   :windows (windows log job-id)
   :triggers (triggers log job-id)
   :metadata (job-metadata log job-id)
   :task-scheduler (rq/task-scheduler replica job-id)})

(defn task-name 
  "Given a task id, returns the task name" 
  [log task-id]
  (:name (extensions/read-chunk log :task task-id)))

(defn task-information
  "Given a job id and task id, returns catalog entry for this task."
  [log job-id task-id]
  (let [task-name (:name (extensions/read-chunk log :task job-id task-id))]
    (find-task (extensions/read-chunk log :catalog job-id) task-name)))


src/onyx/http_query.clj
(ns onyx.http-query
  (:require [onyx.static.default-vals :refer [arg-or-default]]
            [ring.component.jetty :refer [jetty-server]]
            [ring.middleware.params :refer [wrap-params]]
            [com.stuartsierra.component :as component]
            [cheshire.core :refer [generate-string]]
            [clojure.java.jmx :as jmx]
            [onyx.query]
            [onyx.peer-query.job-query :as jq]
            [onyx.system :as system]
            [onyx.peer-query.aeron]
            [onyx.metrics-endpoint :as metrics-endpoint]
            [com.stuartsierra.component :as component]
            [taoensso.timbre :refer [info error infof]]))

(def default-serializer "application/edn")

(def parse-uuid #(java.util.UUID/fromString %))

(defn parse-keyword [kw-str]
  (keyword (clojure.string/replace kw-str #"^:" "")))

(def parsers 
  {:keyword parse-keyword
   :long (fn [s] (if-not (empty? s) (Long/parseLong s)))
   :uuid parse-uuid})

(defn get-param [request param coercer]
  (try 
   ((get parsers coercer) (or (get-in request [:query-params param])
                              (throw (Exception. (str "Missing param " param)))))
   (catch Throwable t
     (throw (Exception. (str "Failed to parse param " param " as " coercer))))))

(defn fetch-from-zookeeper [peer-config f]
  (let [client (component/start (system/onyx-client peer-config))] 
    (try 
     (f (:log client))
     (finally (component/stop client)))))

(defn time-since-heartbeat []
  (when-let [jbean (first (jmx/mbean-names "org.onyxplatform:name=peer-group.since-heartbeat"))]
    (jmx/read jbean "Value")))

(def default-healthy-heartbeat-timeout 40000)

(def endpoints
  {{:uri "/network/media-driver"
    :request-method :get}
   {:doc "Returns a map describing the media driver status."
    :f (fn [request _ _] (onyx.peer-query.aeron/media-driver-health))}

   {:uri "/network/media-driver/active"
    :request-method :get}
   {:doc "Returns a boolean for whether the media driver is healthy and heartbeating."
    :f (fn [request _ _] 
         (let [active (:active (onyx.peer-query.aeron/media-driver-health))]
           {:status (if active 200 500)
            :result active}))}

   {:uri "/health"
    :request-method :get}
   {:doc "Single health check call to check whether the following statuses are healthy: /network/media-driver/active, /peergroup/heartbeat. Call with /health?threshold=30000 for a peer-group heartbeat timeout."
    :f (fn [request _ _] 
         (let [time-since (time-since-heartbeat)
               threshold (or (get-param request "threshold" :long) default-healthy-heartbeat-timeout)
               pg-healthy? (< time-since threshold)
               media-driver-healthy? (:active (onyx.peer-query.aeron/media-driver-health))
               total-healthy? (and pg-healthy? media-driver-healthy?)]
           {:status (if total-healthy? 200 500)
            :result total-healthy?}))}

   {:uri "/peergroup/heartbeat"
    :request-method :get}
   {:doc "Returns the number of milliseconds since the last peer group heartbeat."
    :f (fn [request _ _] 
         {:result (time-since-heartbeat)})}
   
   {:uri "/peergroup/health"
    :request-method :get}
   {:doc "Returns the number of milliseconds since the last peer group heartbeat."
    :f (fn [request _ _] 
         (let [time-since (time-since-heartbeat)
               threshold (or (get-param request "threshold" :long) default-healthy-heartbeat-timeout)
               healthy? (< time-since threshold)]
           {:status (if healthy? 200 500)
            :result healthy?}))}
 
   {:uri "/metrics"
    :request-method :get}
   {:doc "Returns metrics for prometheus"
    :f (fn [request peer-config _]
         {:result (metrics-endpoint/metrics-endpoint peer-config)})}

   {:uri "/replica"
    :request-method :get}
   {:doc "Returns a snapshot of the replica"
    :f (fn [request _ replica]
         {:result replica})}

   {:uri "/replica/peers"
    :request-method :get}
   {:doc "Lists all the peer ids"
    :f (fn [request _ replica]
         {:result (:peers replica)})}

   {:uri "/replica/jobs"
    :request-method :get}
   {:doc "Lists all non-killed, non-completed job ids."
    :f (fn [request _ replica]
         {:result (:jobs replica)})}

   {:uri "/replica/killed-jobs"
    :request-method :get}
   {:doc "Lists all the job ids that have been killed."
    :f (fn [request _ replica]
         {:result (:killed-jobs replica)})}

   {:uri "/replica/completed-jobs"
    :request-method :get}
   {:doc "Lists all the job ids that have been completed."
    :f (fn [request _ replica]
         {:result (:completed-jobs replica)})}

   {:uri "/replica/tasks"
    :request-method :get}
   {:doc "Given a job id, returns all the task ids for this job."
    :query-params-schema {"job-id" String}
    :f (fn [request _ replica]
         (let [job-id (get-param request "job-id" :uuid)]
           {:result (get-in replica [:tasks job-id])}))}

   {:uri "/replica/job-allocations"
    :request-method :get}
   {:doc "Returns a map of job id -> task id -> peer ids, denoting which peers are assigned to which tasks."
    :f (fn [request _ replica]
         {:result (:allocations replica)})}

   {:uri "/replica/task-allocations"
    :request-method :get}
   {:doc "Given a job id, returns a map of task id -> peer ids, denoting which peers are assigned to which tasks for this job only."
    :f (fn [request _ replica]
         (let [job-id (get-param request "job-id" :uuid)]
           {:result (get-in replica [:allocations job-id])}))}

   {:uri "/replica/peer-site"
    :request-method :get}
   {:doc "Given a peer id, returns the Aeron hostname and port that this peer advertises to the rest of the cluster."
    :query-params-schema {"peer-id" String}
    :f (fn [request _ replica]
         (let [peer-id (get-param request "peer-id" :uuid)]
           {:result (get-in replica [:peer-sites peer-id])}))}

   {:uri "/replica/peer-state"
    :request-method :get}
   {:doc "Given a peer id, returns its current execution state (e.g. :idle, :active, etc)."
    :query-params-schema {"peer-id" String}
    :f (fn [request _ replica]
         (let [peer-id (get-param request "peer-id" :uuid)]
           {:result (get-in replica [:peer-state peer-id])}))}

   {:uri "/replica/job-scheduler"
    :request-method :get}
   {:doc "Returns the job scheduler for this tenancy of the cluster."
    :f (fn [request _ replica]
         {:result (:job-scheduler replica)})}

   {:uri "/replica/task-scheduler"
    :request-method :get}
   {:doc "Given a job id, returns the task scheduler for this job."
    :query-params-schema
    {"job-id" String}
    :f (fn [request _ replica]
         {:result (let [job-id (get-param request "job-id" :uuid)]
                    (get-in replica [:task-schedulers job-id]))})}

   {:uri "/job/workflow"
    :request-method :get}
   {:doc (:doc (meta #'jq/workflow))
    :query-params-schema
    {"job-id" String}
    :f (fn [request peer-config replica]
         (fetch-from-zookeeper 
          peer-config
          (fn [log] 
            (let [job-id (get-param request "job-id" :uuid)]
              (jq/workflow log job-id)))))}

   {:uri "/job/catalog"
    :request-method :get}
   {:doc (:doc (meta #'jq/catalog))
    :query-params-schema
    {"job-id" String}
    :f (fn [request peer-config replica]
         {:result (fetch-from-zookeeper 
                   peer-config
                   (fn [log] 
                     (let [job-id (get-param request "job-id" :uuid)]
                       (jq/catalog log job-id))))})}

   {:uri "/job/flow-conditions"
    :request-method :get}
   {:doc (:doc (meta #'jq/flow-conditions))
    :query-params-schema
    {"job-id" String}
    :f (fn [request peer-config replica]
         {:result (fetch-from-zookeeper 
                   peer-config
                   (fn [log] 
                     (let [job-id (get-param request "job-id" :uuid)]
                       (jq/flow-conditions log job-id))))})}

   {:uri "/job/lifecycles"
    :request-method :get}
   {:doc (:doc (meta #'jq/lifecycles))
    :query-params-schema
    {"job-id" String}
    :f (fn [request peer-config replica]
         {:result (fetch-from-zookeeper 
                   peer-config
                   (fn [log] 
                     (let [job-id (get-param request "job-id" :uuid)]
                       (jq/lifecycles log job-id))))})}

   {:uri "/job/windows"
    :request-method :get}
   {:doc (:doc (meta #'jq/windows))
    :query-params-schema
    {"job-id" String}
    :f (fn [request peer-config replica]
         {:result (fetch-from-zookeeper 
                   peer-config
                   (fn [log] 
                     (let [job-id (get-param request "job-id" :uuid)]
                       (jq/windows log job-id))))})}

   {:uri "/job/triggers"
    :request-method :get}
   {:doc (:doc (meta #'jq/triggers))
    :query-params-schema
    {"job-id" String}
    :f (fn [request peer-config replica]
         {:result (fetch-from-zookeeper 
                   peer-config
                   (fn [log] 
                     (let [job-id (get-param request "job-id" :uuid)]
                       (jq/triggers log job-id))))})}

   {:uri "/job/exception"
    :request-method :get}
   {:doc (:doc (meta #'jq/exception))
    :query-params-schema
    {"job-id" String}
    :f (fn [request peer-config replica]
         {:result (fetch-from-zookeeper 
                   peer-config
                   (fn [log] 
                     (let [job-id (get-param request "job-id" :uuid)]
                       (jq/exception log job-id))))})}

   {:uri "/job/task"
    :request-method :get}
   {:doc (:doc (meta #'jq/task-information))
    :query-params-schema
    {"job-id" String
     "task-id" String}
    :f (fn [request peer-config replica]
         {:result (fetch-from-zookeeper 
                   peer-config
                   (fn [log] 
                     (let [job-id (get-param request "job-id" :uuid)
                           task-id (get-param request "task-id" :keyword)]
                       (jq/task-information log job-id task-id))))})}})

(def serializers
  {"application/edn" pr-str
   "application/string" pr-str
   "application/json" generate-string})

(defn ^{:no-doc true} serializer-name
  [content-type]
  (if (serializers content-type)
    content-type
    default-serializer))

(defn ^{:no-doc true} get-serializer
  [content-type]
  (get serializers 
       content-type
       (get serializers default-serializer)))

(defn handler [replica peer-config {:keys [content-type] :as request}]
  (let [serialize (get-serializer content-type)
        f (:f (get endpoints (select-keys request [:request-method :uri])))]
    (try 
     (if-not f
       {:status 404
        :headers {"Content-Type" (serializer-name content-type)}
        :body (serialize {:status :failed :message "Endpoint not found."})}
       (let [{:keys [status result]} (f request peer-config @replica)]
         {:status (or status 200)
          :headers {"Content-Type" (serializer-name content-type)}
          :body (if (= "/metrics" (:uri request)) 
                  result
                  (serialize {:status :success
                              :result result}))}))
     (catch Throwable t
       (error t "HTTP peer health query error")
       {:status 500
        :body (pr-str t)}))))

(defn app [replica peer-config]
  {:handler (wrap-params (fn [request] (handler replica peer-config request)))})

(defrecord QueryServer [replica server peer-config]
  component/Lifecycle
  (start [this]
    (let [ip (arg-or-default :onyx.query.server/ip peer-config)
          port (arg-or-default :onyx.query.server/port peer-config)
          replica (atom {})
          server-component (jetty-server {:app (app replica peer-config) :host ip :port port})]
      (infof "Starting http query server on %s:%s" ip port)
      (assoc this 
             :replica replica
             :server (component/start server-component))))
  (stop [this]
    (info "Stopping http query server")
    (assoc this 
           :replica nil
           :server (component/stop server)
           :loggin-config nil
           :peer-config nil)))

(defmethod onyx.query/query-server true [peer-config]
  (map->QueryServer {:peer-config peer-config}))

(defmethod clojure.core/print-method QueryServer
  [system ^java.io.Writer writer]
  (.write writer "#<Query Server>"))
